Introduce the advanced sampling notebook into PySAGES (#165)
SSAGESLabs/PySAGES: examples/Advanced_Sampling_Introduction.md

Ludwig Schneider and Juan de Pablo

Pritzker School of Molecular Engineering

The University of Chicago

Berlin, July 28th, 2022
<!-- #endregion -->

<!-- #region id="T-Qkg9C9n7Cc" -->
# Setting up the environment

First, we are setting up our environment. We use an already compiled and packaged installation of HOOMD-blue and the DLExt plugin.
We copy it from Google Drive and install PySAGES for it.
<!-- #endregion -->

```bash id="3eTbKklCnyd_"

BASE_URL="https://drive.google.com/u/0/uc?id=1hsKkKtdxZTVfHKgqVF6qV2e-4SShmhr7&export=download"
wget -q --load-cookies /tmp/cookies.txt "$BASE_URL&confirm=$(wget -q --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate $BASE_URL -O- | sed -rn 's/.*confirm=(\w+).*/\1\n/p')" -O pysages-env.zip
rm -rf /tmp/cookies.txt
```

```python colab={"base_uri": "https://localhost:8080/"} id="KRPmkpd9n_NG" outputId="04c4281e-57d3-4f48-ce52-fb010a585708"
%env PYSAGES_ENV=/env/pysages
```

```bash id="J7OY5K9VoBBh"

mkdir -p $PYSAGES_ENV .
unzip -qquo pysages-env.zip -d $PYSAGES_ENV
```

```bash colab={"base_uri": "https://localhost:8080/"} id="LlVSU_-FoD4w" outputId="85971918-c9b9-44b6-8898-3557ae165cc9"

> apt-cache policy libcudnn8
> apt install --allow-change-held-packages libcudnn8=8.4.1.50-1+cuda11.6

Reviewer: ```suggestion
apt-cache policy libcudnn8 &> /dev/null
apt install --allow-change-held-packages libcudnn8=8.4.1.50-1+cuda11.6 &> /dev/null
```
I would make the output less noisy here.

```

<!-- #region id="we_mTkFioS6R" -->
## PySAGES

The next step is to install PySAGES.
First, we install the jaxlib version that matches the CUDA installation of this Colab setup. See the JAX documentation [here](https://github.com/google/jax) for more details.
<!-- #endregion -->

```bash id="vK0RZtbroQWe"

pip install -q --upgrade pip
# Installs the wheel compatible with CUDA 11 and cuDNN 8.2 or newer.
pip install -q --upgrade "jax[cuda]" -f https://storage.googleapis.com/jax-releases/jax_cuda_releases.html &> /dev/null
```

<!-- #region id="wAtjM-IroYX8" -->
Now we can finally install PySAGES. We clone the newest version from [here](https://github.com/SSAGESLabs/PySAGES) and build the remaining pure python dependencies and PySAGES itself.
<!-- #endregion -->

```bash id="B-HB9CzioV5j"

rm -rf PySAGES
git clone https://github.com/SSAGESLabs/PySAGES.git &> /dev/null
cd PySAGES
pip install -q . &> /dev/null
```

<!-- #region id="KBFVcG1FoeMq" -->
# Advanced Sampling Methods
<!-- #endregion -->

```bash id="ppTzMmyyobHB"

mkdir /content/advanced_sampling
cd /content/advanced_sampling
```

<!-- #region id="jnMtDUgyXEsr" -->

> Many systems have a separated free-energy landscape, where different basins are divided by energy barriers.

Reviewer: ```suggestion
Many systems have rugged free-energy landscapes, where different basins are separated by energy barriers.
```

### Examples:
#### Collective molecular transitions
* Protein folding
* Protein ligand binding
* Liqid membrane fusion
<!-- #endregion -->

<!-- #region id="mHOcKW-fXoK5" -->

> <img src=https://pubs.acs.org/cms/10.1021/ja910050x/asset/images/large/ja-2009-10050x_0008.jpeg width=500>
> 
> <img src=https://pubs.acs.org/cms/10.1021/ja910050x/asset/images/large/ja-2009-10050x_0011.jpeg width=500>

Reviewer: ```suggestion
<div>
  <img src=https://pubs.acs.org/cms/10.1021/ja910050x/asset/images/large/ja-2009-10050x_0008.jpeg width=500>
</div>

<div>
  <img src=https://pubs.acs.org/cms/10.1021/ja910050x/asset/images/large/ja-2009-10050x_0011.jpeg width=500>
</div>
```
Display them one on top of the other?

Smirnova, Yuliya G., et al. "Solvent-exposed tails as prestalk transition states for membrane fusion at low hydration." Journal of the American Chemical Society 132.19 (2010): 6710-6718
<!-- #endregion -->

<!-- #region id="hz7pwhbZXzZp" -->
#### Phase transisions
* Liquid-vapor transisition
* Crystalization
* Shape transitions
* Magnetization (Ising model)

> ![picture](https://upload.wikimedia.org/wikipedia/commons/e/e6/Ising_quench_b10.gif)
> Wikipedia: https://en.wikipedia.org/wiki/Ising_model

Reviewer: ```suggestion
![picture](https://upload.wikimedia.org/wikipedia/commons/e/e6/Ising_quench_b10.gif)

Wikipedia: https://en.wikipedia.org/wiki/Ising_model
```
I think this would be better displayed this way

<!-- #endregion -->

<!-- #region id="0iCFnwF6aJ3P" -->

> ###Collective Variable (order parameter)

Reviewer: ```suggestion
### Collective Variables

Also known as order parameters or reaction coordinates in other contexts.
```
Just a suggestion in case we might want to be more thorough.

### Collective Variables
Also known as order parameters in other contexts.

Usually, such transitions can be described in terms of order parameters/collective variables (CV).

$⇒$ Reduction of the phase space $\hat\xi(\{\vec{r}, \vec{p}\})$. In general, a collective variable can be multi-dimensional.
[Note: the hat "$\hat\ $" indicates that it is a collective variable that explicitly depends on the underlying molecular configuration. Without the "$\hat\ $" it usually indicates a scalar, like a target.]

Examples:
* center of mass of the molecule
* opening angle of protein
* magnetization
* [more common options in PySAGES](https://pysages.readthedocs.io/en/latest/module-pysages-colvars.html)
<!-- #endregion -->

<!-- #region id="0W2ukJuuojAl" -->
## Example system: particle in multi-well

Let's start exploring an easy system and find the free-energy landscape.

* a priori $A(\xi)$ is unknown
* but we can measure the density of states $p(\xi)$

We can recover the free-energy profile from $p(\xi)$

$$A = - \frac{1}{\beta}\ln(Z)$$

* $A$: free-energy
* $Z$: NVT partition function

We can formulate the partial partition function for a collective variable $\xi$

> $$Z(\xi) = \int\text{d}^N\vec{r}_i\text{d}^N\vec{p}_i \delta (\hat\xi(\{\vec{r}, \vec{p}\} - \xi) e^{-\beta H(\{(\vec{r}_i, \vec{p}_i\})}$$

Reviewer 2: H**(**\{(\vec{r}_i, \vec{p}_i\})}$$ to H\{(\vec{r}_i, \vec{p}_i\})}$$

Normalization:
$$p(\xi) = \frac{\int\text{d}^N\vec{r}_i\text{d}^N\vec{p}_i \delta (\hat\xi(\{\vec{r}, \vec{p}\} -\xi) e^{-\beta H(\{(\vec{r}_i, \vec{p}_i\})}}{\int\text{d}^N\vec{r}_i\text{d}^N\vec{p}_i e^{-\beta H(\{(\vec{r}_i, \vec{p}_i\})}}$$

$$\Rightarrow A(\xi) = -\frac{1}{\beta} \ln(p(\xi)) + C$$
We can determine the free-energy profile $A(\xi)$ from the probability distribution $p(\xi)$.

With regular MC/MD a finite simulation trajectory is confined to a single basin as the probability of leaving the basin scales as

$$p \propto e^{-\beta \Delta E} \ll 1$$

> 

Reviewer 2: where $$ \Delta E $$ is the energy required to leave the basin?

> 

Reviewer 2: This can or This implies?

* system might not equilibrate

> * trajectory is not ergodic

Reviewer 2: system is not ergodic? though, we assume it to be ergodic.
